src/storage/src/hummock/shared_buffer/shared_buffer_manager.rs

```diff
@@ -99,16 +206,61 @@ impl SharedBufferManager {
         Ok(())
     }
 
+    async fn flush(&self) -> HummockResult<()> {
+        assert!(!self.empty());
+        let mut res = Ok(());
+
+        let notifier = self.ongoing_flush.subscribe();
+        if let Some(mut rx) = notifier {
+            log::debug!("flush: wait for notification");
+            // Wait for notification
+            rx.changed().await.unwrap();
+            if !(*rx.borrow()) {
+                res = Err(HummockError::shared_buffer_error(
+                    "Fail to flush shared buffer",
+                ));
+            }
+        } else {
+            self.ongoing_flush.init();
```

Comment:
What if there're two coroutines running here at the same time? Seems this will panic on assertion. The `subscribe` and `init` should be executed atomically.

Reply:
That's a problem.